### ISSUE TITLE
ClusterAPI is deleting machines after cluster-nodes cannot be retrieved (Unauthorized)

### ISSUE THREAD

**Issue by xvzf** [kubernetes-sigs/cluster-api#5914]
**What steps did you take and what happened:**
With the cluster-api-provider-aws, the `kubeconfig` secret contains a hinted token which is retrieved by the AWS STS service (check [here](https://github.com/kubernetes-sigs/aws-iam-authenticator#api-authorization-from-outside-a-cluster) for reference) and needs to be refreshed every 15 minutes in the [current configuration](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/1a61dfc39649749b8b8dfad8334909a388d675a7/pkg/cloud/services/eks/config.go#L261).

Instead of recreating the client for the observed cluster before the token expires/after the token expires, cluster-api fails with unauthorised errors and starts deleting/recreating all machines.
I couldn't find an explanation based on the code but it can be reproduced with confidence in our setup.

**What did you expect to happen:**
The client is recreated *and* after an auth failure nodes are not arbitrarily deleted.

**Anything else you would like to add:**
Logs of all `cap*` component: https://gist.github.com/xvzf/78d47ce6c3d6fabd49bc902e5d22d467
Manifests to reproduce it: https://gist.github.com/xvzf/87c934945ad97fe3bb9ee4934c6478ce

*Note:* Interestingly this is also affecting the cluster-autoscaler, failing unauthorised, restarting and continuing. It might also be a bug in the `client-go` dependency.

**Environment:**

- Cluster-api version: v1.0.2
- Minikube/KIND version: EKS 1.21.2 (should be irrelevant)
- Kubernetes version: (use `kubectl version`): 1.21.2
- OS (e.g. from `/etc/os-release`): AmazonLinux

/kind bug
/area health
/area provider/aws
[One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels]

**Comment by xvzf** [comment]
Another thing we realised: the `machinedeployment` (or the `machineset`) replica count is updated when a `machine` is deleted by cluster-api. Is this intended?

**Comment by fabriziopandini** [comment]
@xvzf probably it is better to open this issue in cluster-api-provider-aws repository, so the team can triage the AWS specific bits first...

**Comment by xvzf** [comment]
@fabriziopandini we weren't sure where to open this issue, I'll create an issue on the aws provider as well and reference this one.

The deleting and implicit scaling down of the `machinedeployment` belongs here. It might be worth splitting this up into two issues though

**Comment by enxebre** [comment]
Hey @xvzf I haven't had time to debug this yet, but

>Another thing we realised: the machinedeployment (or the machineset) replica count is updated when a machine is deleted by cluster-api. Is this intended?

I can't think of any core controller that would modify machineDeployment.spec.replicas. So as a first thought this is possibly a red herring. Have you seen that happening in a bare cluster (no autoscaler, no flux...)?

**Comment by xvzf** [comment]
@enxebre 
> I can't think of any core controller that would modify machineDeployment.spec.replicas. So as a first thought this is possibly a red herring. Have you seen that happening in a bare cluster (no autoscaler, no flux...)?

Yes, both of the mentioned components scaled down to 0

**Comment by sbueringer** [comment]
> @enxebre
> 
> > I can't think of any core controller that would modify machineDeployment.spec.replicas. So as a first thought this is possibly a red herring. Have you seen that happening in a bare cluster (no autoscaler, no flux...)?
> 
> Yes, both of the mentioned components scaled down to 0

I could be wrong, but I couldn't fine a single write operation on MachineDeployment.spec.replicas in the regular controllers (there is one in the controller for ClusterClass, but I think ClusterClass is not used here).

To be absolutely sure we have the right data. Where are you "reading" the replicas of the MachineDeployment from? The "REPLICAS" column of the MachineDeployment is `.status.replicas" not `.spec.replicas".

**Comment by xvzf** [comment]
@sbueringer Yes, it's been the `.spec.replicas`; I stepped through the code and couldn't find anything as well (just updating the specs). I'll try to check with audit logging which service account is doing the change, this way we might habe a better clue on what's going on.

What's strange is the ordered downscaling from `4 -> 3 -> 2 -> ...` so there's likely _something_ that performs those actions.

**Comment by sbueringer** [comment]
If you refer to the events, it looks to me like something scales down the MD.spec.replicas and then the MD controller scales down the MachineSet accordingly.

**Comment by xvzf** [comment]
The root cause of the node deletion has been an assumingly terminated instance of cluster-autoscaler with a different configuration (it was still up in containerd).

Thx for the help folks!
